### DISCUSSION TITLE
point cloud stream

### DISCUSSION THREAD

**Discussion by Jaetriel** [lucasw/rviz_camera_stream#35] [Q&A]
Hey guys, is it possible to use this package to publish a sort of 'captured image' of a point cloud topic and publish those?

**Comment by flynneva** [comment]
@Jaetriel the recently published "rviz_animated_view_controller" does this sort of by publishing an image topic of the current rviz view if that's what you are looking for.

**Comment by lucasw** [comment]
If rviz can view it through the main window then rviz_camera_stream should capture it (and I imagine `rviz_animated_view_controller` is doing something similar with ogre and can do it too, but is limited to only the rviz main view?).

Because rviz_camera_stream can be attached to any tf frame with very different perspectives than the main view there can be strange view dependent artifacts- Markers or billboard rendering styles that align to the main rviz view will be seen at an angle here, and lighting will be funny because there is an rviz main view light source- but if point cloud points are set to render as points and aren't lit by that light it should be fine.  (There is an rviz plugin that allows creation of other light sources, I think attaching them to tf frames, I'd have to dig it up)

**Reply by Jaetriel** [reply]
I did manage to somehwat get it to work, and it can show the point cloud, I'm just struggling to attach it to the correct odometry frame which shows the robot moving to achieve the effect of the camera "following" the robot as it's moving through and creating the map but that's mostly probably due to my poor knowledge and skills when it comes to URDFs and frames in ROS

**Reply by lucasw** [reply]
You may be better off with a static_transform_publisher (http://wiki.ros.org/tf#static_transform_publisher) instead of modifying your urdf, it could be a transform from your base_link to your virtual following camera frame, this for instance puts a camera 2 meters behind and 1 meter above the base_link with a little downward tilt:

```
  <node pkg="tf" type="static_transform_publisher" name="follow_camera_broadcaster"
      args="-2.0 0.0 1.0 0 0.2 0 base_link follow_camera1 10" />
```

If you are free to share a screenshot that could be helpful.

**Reply by Jaetriel** [reply]
Hey, sorry for the late answer, so, the tf tree normally looks like this:
![frames_normal](https://user-images.githubusercontent.com/43788971/134341755-8d619c87-a408-4e97-a13e-f2abb16a2517.png)
I tried attaching a static_transform_publisher from `body` ( where the Odometry is published, it's the frame that in rviz shows the robot moving through the map, and `camera_init` is the frame where the point cloud map is published  ) to the camera like so:
![frames_body](https://user-images.githubusercontent.com/43788971/134344824-dedce480-1ca1-443c-a498-5ec009d9a9fa.png)
but for some reason doing it like this I'm not seeing anything. I then tried to make the transform publisher from `camera_init`(lidar frame)
to camera1 and I can see the point cloud, but the camera does not move obviously, so I feel like I am missing something. Here's a picture of the whole graph as well maybe it can shed some more light:
![rosgraph_1](https://user-images.githubusercontent.com/43788971/134345280-59da907b-6492-4347-b0b1-054b30aa3746.png)

**Reply by lucasw** [reply]
It looks like you are playing the `laserMapping` from a bag?  That's causing the transform out to be stale - the first two images shows the most recent `body` transform is 1026991 seconds old, so tf lookups (that rviz_camera_stream and many other nodes depend on) won't work if the tf lookup involves that old frame.

You want to play back the bag data using `--clock` and set `use_sim_time` to true, launch the live parts of your system after that, and then everything will be operating as if back in time when the bag was captured.

**Reply by Jaetriel** [reply]
Thank you that actually worked! Now I just need to find a way to properly position it, I want to achieve basically the same as having a XYOrbit type of view, and I tried using the tf_publisher_gui but it doesn't work well when I have the `static_transform_publisher` on and when I dont, the camera frame doesn't get published for some reason.

**Reply by lucasw** [reply]
The static transform publisher and tf publisher gui are probably conflicting with each other if they are trying to broadcast the same child frame id, but not sure why running just the tf publisher gui wouldn't work.

It doesn't do XY orbit (at least not as smoothly mouse controlled as the rviz camera) but this is my contribution to the interactive setting of tfs: https://github.com/lucasw/rviz_interactive_tf - and there are a few more from others with different approaches- you may want to fork one of those and modify it to get exactly what you want (though I don't think you can intercept raw mouse moves in rviz outside of an rviz view controller), given you can get it working at the basic level first.

**Reply by Jaetriel** [reply]
Soo my idea of creating a sort of "interactive" camera image being published might turn out more difficult than I thought. Basically my goal in the end would be to output in real-time the map generated by the SLAM and I thought maybe I can try and transfer it over rosbridge as images captured by a virtual camera in rviz instead of point cloud topic messages because those are a lot bigger and probably take a lot more resources to be processed and rendered on the client side, but I would still really like to have the interactivity of being able to turn/pan/zoom the view like you can in rviz

**Reply by lucasw** [reply]
If you do keep pursuing this you may be better off skipping rviz as an intermediate step, have something subscribe to the map or point cloud and do some operations to get it into the format of your choice, send that out. 

I'd expect that the slam process is already publishing an occupancy grid 2D map, and that is much less cumbersome than a point cloud and can be output to rosbridge easily (maybe you have to convert it into a regular Image, which shouldn't be too hard).

**Reply by Jaetriel** [reply]
Thanks for the suggestion, I might try and see if I can get good enough performance just rendering the point cloud on the client-side itself using something like ros3djs
